include/kframework/lemmas/bytes-simplification.k

```diff
@@ -68,71 +68,159 @@ module BYTES-SIMPLIFICATION [symbolic]
+      #range(B1:Bytes +Bytes B2:Bytes, S:Int, W:Int) => #range(#range(B1, S, lengthBytes(B1) -Int S) +Bytes B2, 0, W)
+      requires 0 <Int S andBool S <=Int lengthBytes(B1)
+      [simplification]
+
+    rule [range-of-range]:
+      #range(#range(B:Bytes, S1:Int, W1:Int), S2:Int, W2:Int) =>
+        #range(B, S1 +Int S2, minInt(W2, W1 -Int S2)) +Bytes #buf(maxInt(0, W2 -Int (W1 -Int S2)), 0)
+      requires 0 <=Int S1 andBool 0 <=Int W1 andBool 0 <=Int S2 andBool 0 <=Int W2
+      [simplification]
 
-    rule #range( M [ N := BUF:Bytes ], L, K ) =>
-           #let S = L -Int N #in
-           #let W = minInt(K, lengthBytes(BUF) -Int S) #in
-           #range(BUF, S, W) +Bytes #range(M, N +Int lengthBytes(BUF), K -Int W)
-      requires 0 <Int K andBool N <=Int L andBool L <=Int N +Int lengthBytes(BUF) [simplification]
+    rule [range-join-base]:
+      #range(B, S1, W1) +Bytes #range(B, S2, W2) => #range(B, S1, W1 +Int W2)
+      requires S2 ==Int S1 +Int W1
+       andBool 0 <=Int S1 andBool 0 <=Int W1 andBool 0 <=Int S2 andBool 0 <=Int W2
+      [simplification]
 
-    rule #range( M [ N := BUF:Bytes ], L, K ) => #range(M, L, K)
-      requires 0 <Int K andBool N +Int lengthBytes(BUF) <=Int L [simplification]
+    rule [range-join-concat]:
+      #range(B1, S1, W1) +Bytes ( #range(B1, S2, W2) +Bytes B2 ) => #range(B1, S1, W1 +Int W2) +Bytes B2
+      requires S2 ==Int S1 +Int W1
+       andBool 0 <=Int S1 andBool 0 <=Int W1 andBool 0 <=Int S2 andBool 0 <=Int W2
+      [simplification]
 
-    rule #range( M [ N1 := BUF1:Bytes ] [ N2 := BUF2:Bytes ], L, K ) => #range( M [ N2 := BUF2 ], L, K )
-        requires 0 <Int K andBool N1 +Int lengthBytes(BUF1) <=Int L [simplification]
+    rule [range-reform-base]:
+      #range(B:Bytes, S:Int, minInt(W1:Int, W2:Int)) +Bytes #buf(maxInt(0, W3:Int), 0) => #range (B, S, W1)
+      requires W2 ==Int lengthBytes(B) -Int S
+       andBool W3 ==Int S +Int W1 -Int lengthBytes(B)
+       andBool 0 <=Int S andBool 0 <=Int W1 andBool 0 <=Int W2
+      [simplification]
 
-    rule #range(M, N1 +Int N2, K) => #range(#range(M, N2, lengthBytes(M) -Int N2), N1, K)
-      requires 0 <=Int N1 andBool 0 <=Int N2
-      [simplification, concrete(N2), concrete(M)]
+    rule [range-reform-concat]:
+      #range(B1:Bytes, S:Int, minInt(W1:Int, W2:Int)) +Bytes ( #buf(maxInt(0, W3:Int), 0) +Bytes B2:Bytes ) => #range (B1, S, W1) +Bytes B2
+      requires W2 ==Int lengthBytes(B1) -Int S
+       andBool W3 ==Int S +Int W1 -Int lengthBytes(B1)
+       andBool 0 <=Int S andBool 0 <=Int W1 andBool 0 <=Int W2
+      [simplification]
 
-    rule #range(A +Bytes _B, START, LENGTH) => #range(A, START, LENGTH)
-      requires START +Int LENGTH <=Int lengthBytes(A) [simplification]
+    rule [range-buf-zero-1]:
+      #range(#buf(32, X:Int), 0, W:Int) => #buf(W, 0)
+      requires 0 <=Int W andBool W <=Int 32
+       andBool #rangeUInt (8 *Int (32 -Int W), X)
+      [simplification, concrete(W)]
```

Comment:
Are we sure this rule is doing anything?

`#rangeUInt(8 *Int (32 -Int W), X)` probably actually doesn't get understood by the backend, because `#rangeUInt(...)` is an `alias` (which is a `macro`), so to th ebackend it's probably just an uninterpreted symbol (and likely leads to it being very confused). Macros (and aliases) are expanded at kompile time, and the backend should not see the rules for it. Does removing this lemma break anything?

If it is needed, I think it should be woreded instead as:

```suggestion
    rule [range-buf-zero-1]:
      #range(#buf(32, X:Int), 0, W:Int) => #buf(W, 0)
      requires 0 <=Int W andBool W <=Int 32
       andBool 0 <=Int X andBoool X <Int 8 *Int (32 -Int W)
      [simplification, concrete(W)]
```

Reply:
Is this actually subsumed completely by `range-buf-zero` further down?

Reply:
Yes, good point - but `X <Int 2 ^Int (8 *Int (32 -Int W))`, right?

It is subsumed completely, yes, by `range-buf-zero`.